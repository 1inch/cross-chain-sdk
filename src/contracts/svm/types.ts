import {SolanaAddress} from '../../domains'

export type CreateOrderAccounts = {
    creator: SolanaAddress
    mint: SolanaAddress
    creatorAta: SolanaAddress
    order: SolanaAddress
    orderAta: SolanaAddress
    associatedTokenProgram: SolanaAddress
    tokenProgram: SolanaAddress
    rent: SolanaAddress
    systemProgram: SolanaAddress
}
