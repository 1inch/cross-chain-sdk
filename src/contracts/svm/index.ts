export {SvmDstEscrowFactory} from './svm-dst-escrow-factory.js'
export {SvmSrcEscrowFactory} from './svm-src-escrow-factory.js'
export {WhitelistContract as SvmWhitelistContract} from './whitelist.js'
export {
    Instruction as SvmInstruction,
    AccountMeta as SvmAccountMeta
} from './instruction.js'

export {
    CreateOrderAccounts,
    ParsedCreateDstEscrowInstructionData,
    ParsedCreateInstructionData,
    ParsedCreateSrcEscrowInstructionData
} from './types.js'
