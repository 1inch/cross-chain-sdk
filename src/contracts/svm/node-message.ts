export type NodeMessage = {
    accountKeys: {
        pubkey: string
        signer: boolean
        source: 'transaction' | 'lookupTable'
        writable: boolean
    }[]
    instructions: {
        accounts: string[]
        data: string // base58 encoded
        programId: string
    }[]
    addressTableLookups?: {
        accountKey: string
        writableIndexes: number[]
        readonlyIndexes: number[]
    }[]
}
