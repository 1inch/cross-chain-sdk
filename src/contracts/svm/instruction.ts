import bs58 from 'bs58'
import {Buffer} from 'buffer'
import {NodeMessage} from './node-message'
import {SolanaAddress} from '../../domains/addresses'

export class Instruction {
    constructor(
        /**
         * Program Id to execute
         */
        public readonly programId: SolanaAddress,
        public readonly accounts: Array<AccountMeta>,

        /**
         * Program input
         */
        public readonly data: Buffer
    ) {}

    static fromNode(msg: NodeMessage): Instruction[] {
        return msg.instructions.map((ix) => {
            return new Instruction(
                new SolanaAddress(ix.programId),
                ix.accounts.map((pubkey) => {
                    const account = msg.accountKeys.find(
                        (x) => x.pubkey === pubkey
                    )

                    if (!account) {
                        throw new Error('account not found')
                    }

                    return {
                        isWritable: account.writable,
                        isSigner: account.signer,
                        pubkey: new SolanaAddress(account.pubkey)
                    }
                }),
                Buffer.from(bs58.decode(ix.data))
            )
        })
    }
}

export type AccountMeta = {
    /** An account's public key */
    pubkey: SolanaAddress
    /** True if an instruction requires a transaction signature matching `pubkey` */
    isSigner: boolean
    /** True if the `pubkey` can be loaded as a read-write account. */
    isWritable: boolean
}
