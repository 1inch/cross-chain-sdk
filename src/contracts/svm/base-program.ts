import {AccountMeta} from './instruction'
import {SolanaAddress} from '../../domains/addresses'

export abstract class BaseProgram {
    protected encoder = new TextEncoder()

    constructor(public readonly programId: SolanaAddress) {}

    protected optionalAccount(meta: AccountMeta, skip: boolean): AccountMeta {
        if (skip) {
            return {
                pubkey: this.programId,
                isSigner: false,
                isWritable: false
            }
        }

        return meta
    }
}
