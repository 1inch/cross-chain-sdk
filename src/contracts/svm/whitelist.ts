import {BaseProgram} from './base-program'
import {SolanaAddress} from '../../domains/addresses'
import {getPda} from '../../utils'

export class WhitelistContract extends BaseProgram {
    static DEFAULT = new WhitelistContract(
        new SolanaAddress('5XYZ3LMWECpC6u7BWLskMMNx4xWbXF44dpDxvkVqkHtA')
    )

    constructor(programId: SolanaAddress) {
        super(programId)
    }

    public getAccessAccount(taker: SolanaAddress): SolanaAddress {
        return getPda(this.programId, [
            this.encoder.encode('resolver_access'),
            taker.toBuffer()
        ])
    }
}
