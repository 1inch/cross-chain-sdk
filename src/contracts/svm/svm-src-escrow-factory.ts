import {BN, BorshCoder} from '@coral-xyz/anchor'
import {Instruction} from './instruction'
import {BaseProgram} from './base-program'
import {WhitelistContract} from './whitelist'
import {HashLock, MerkleLeaf, SolanaAddress} from '../../domains'
import {getAta} from '../../utils'
import {SvmCrossChainOrder} from '../../cross-chain-order/svm/svm-cross-chain-order'
import {IDL} from '../../idl/cross-chain-escrow-src'
import {uint256split} from '../../utils/numbers/uint256-split'
import {hashForSolana} from '../../domains/auction-details/hasher'

export class SvmSrcEscrowFactory extends BaseProgram {
    static DEFAULT = new SvmSrcEscrowFactory(
        new SolanaAddress('2g4JDRMD7G3dK1PHmCnDAycKzd6e5sdhxqGBbs264zwz')
    )

    private readonly coder = new BorshCoder(IDL)

    constructor(programId: SolanaAddress) {
        super(programId)
    }

    public createOrder(
        order: SvmCrossChainOrder,
        extra: {
            srcTokenProgramId: SolanaAddress
        }
    ): Instruction {
        const data = this.coder.instruction.encode('create', {
            hashlock: order.hashLock.toBuffer(),
            amount: new BN(order.makingAmount.toString()),
            safetyDeposit: new BN(order.srcSafetyDeposit.toString()),
            timelocks: uint256split(order.timeLocks.build()).map(bigIntToBN),
            expirationTime: Number(order.deadline),
            assetIsNative: order.srcAssetIsNative,
            dstAmount: uint256split(order.takingAmount).map(bigIntToBN),
            dutchAuctionDataHash: hashForSolana(order.auction),
            maxCancellationPremium: new BN(
                order.resolverCancellationConfig.maxCancellationPremium.toString()
            ),
            cancellationAuctionDuration:
                order.resolverCancellationConfig.cancellationAuctionDuration,
            allowMultipleFills: order.multipleFillsAllowed,
            salt: new BN(order.salt.toString()),
            _dstChainParams: {
                chainId: order.dstChainId,
                makerAddress: order.receiver.toBuffer(),
                token: order.takerAsset.toBuffer(),
                safetyDeposit: new BN(order.srcSafetyDeposit.toString())
            }
        })

        const orderAccount = order.getOrderAccount(this.programId)

        return new Instruction(
            this.programId,
            [
                // 1. maker
                {pubkey: order.maker, isWritable: true, isSigner: true},
                // 2. src mint
                {
                    pubkey: order.makerAsset,
                    isWritable: false,
                    isSigner: false
                },
                // 3. src_maker_ata
                this.optionalAccount(
                    {
                        pubkey: getAta(
                            order.maker,
                            order.makerAsset,
                            extra.srcTokenProgramId
                        ),
                        isSigner: false,
                        isWritable: true
                    },
                    order.srcAssetIsNative
                ),
                // 4. order
                {
                    pubkey: orderAccount,
                    isSigner: false,
                    isWritable: true
                },
                // 5. order_ata
                {
                    pubkey: getAta(
                        orderAccount,
                        order.makerAsset,
                        extra.srcTokenProgramId
                    ),
                    isSigner: false,
                    isWritable: true
                },
                // 6. associated_token_program
                {
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                // 7. token_program
                {
                    pubkey: extra.srcTokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                // 8. rent
                {
                    pubkey: SolanaAddress.SYSVAR_RENT_ID,
                    isSigner: false,
                    isWritable: false
                },
                // 9. system_program
                {
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }

    public createEscrow(
        order: SvmCrossChainOrder,
        fillAmount: bigint,
        extra: {
            /**
             * Who will sign tx
             */
            taker: SolanaAddress
            /**
             * If not passed, than `WhitelistContract.DEFAULT` will be used
             * @see WhitelistContract.DEFAULT
             */
            whitelistProgramId?: SolanaAddress
            /**
             * TokenProgram or TokenProgram 2022
             */
            srcTokenProgramId: SolanaAddress
            /**
             * Required if order allows partial fills
             */
            merkleProof?: {
                /**
                 * Merkle proof for index `idx`
                 *
                 * @see HashLock.getProof
                 */
                proof: MerkleLeaf[]
                /**
                 * @see SvmCrossChainOrder.getMultipleFillIdx
                 */
                idx: number
                /**
                 * Hash of secret at index `idx`
                 */
                secretHash: Buffer
            }
        }
    ): Instruction {
        const {merkleProof} = extra
        const whitelistProgram = extra.whitelistProgramId
            ? new WhitelistContract(extra.whitelistProgramId)
            : WhitelistContract.DEFAULT

        const data = this.coder.instruction.encode('createEscrow', {
            fillAmount: new BN(fillAmount.toString()),
            auction: {
                startTime: Number(order.auction.startTime),
                duration: Number(order.auction.duration),
                initialRateBump: Number(order.auction.initialRateBump),
                pointsAndTimeDeltas: order.auction.points.map((p) => ({
                    rateBump: p.coefficient,
                    timeDelta: p.delay
                }))
            },
            merkleProof: merkleProof && {
                proof: merkleProof.proof.map((p) => Buffer.from(p.slice(2))),
                idx: new BN(merkleProof.idx),
                secretHash: merkleProof.secretHash
            }
        })

        const orderAccount = order.getOrderAccount(this.programId)
        const escrowAddress = order.getEscrowAddress(
            this.programId,
            extra.taker,
            merkleProof?.secretHash
        )

        return new Instruction(
            this.programId,
            [
                // 1. taker
                {
                    pubkey: extra.taker,
                    isSigner: true,
                    isWritable: true
                },
                // 2. resolver_access
                {
                    pubkey: whitelistProgram.getAccessAccount(extra.taker),
                    isSigner: false,
                    isWritable: false
                },
                // 3. maker
                {
                    pubkey: order.maker,
                    isSigner: false,
                    isWritable: true
                },
                // 4. mint
                {
                    pubkey: order.makerAsset,
                    isSigner: false,
                    isWritable: false
                },
                // 5. order
                {
                    pubkey: orderAccount,
                    isSigner: false,
                    isWritable: true
                },
                // 6. order ata
                {
                    pubkey: getAta(
                        orderAccount,
                        order.makerAsset,
                        extra.srcTokenProgramId
                    ),
                    isSigner: false,
                    isWritable: true
                },
                // 7. escrow
                {
                    pubkey: escrowAddress,
                    isSigner: false,
                    isWritable: true
                },
                // 8. escrow ata
                {
                    pubkey: getAta(
                        escrowAddress,
                        order.makerAsset,
                        extra.srcTokenProgramId
                    ),
                    isSigner: false,
                    isWritable: true
                },
                // 9. associated_token_program
                {
                    pubkey: SolanaAddress.ASSOCIATED_TOKE_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                },
                // 10. token_program
                {
                    pubkey: extra.srcTokenProgramId,
                    isSigner: false,
                    isWritable: false
                },
                // 11. system_program
                {
                    pubkey: SolanaAddress.SYSTEM_PROGRAM_ID,
                    isSigner: false,
                    isWritable: false
                }
            ],
            data
        )
    }
}

function bigIntToBN(i: bigint): BN {
    return new BN(i.toString())
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type _ = HashLock // to have ability to refer to it in jsdoc
