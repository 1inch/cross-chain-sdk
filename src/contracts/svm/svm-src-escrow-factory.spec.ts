import {parseEther, parseUnits} from 'ethers'
import {Buffer} from 'buffer'
import {randomBytes} from 'crypto'
import {SvmSrcEscrowFactory} from './svm-src-escrow-factory'
import {NetworkEnum} from '../../chains'
import {SvmCrossChainOrder} from '../../cross-chain-order/svm/svm-cross-chain-order'
import {EvmAddress, SolanaAddress} from '../../domains/addresses'
import {HashLock} from '../../domains/hash-lock'
import {TimeLocks} from '../../domains/time-locks'
import {AuctionDetails} from '../../domains/auction-details'
import {ResolverCancellationConfig} from '../../cross-chain-order'
import {Immutables} from '../../domains/immutables'
import {bufferFromHex} from '../../utils/bytes'

describe('SVM Escrow src factory', () => {
    it('should generate create instruction from order', () => {
        const order = SvmCrossChainOrder.new(
            // 1 WETH [solana] -> 1000 USDC [ethereum]
            {
                maker: SolanaAddress.fromBigInt(1n),
                receiver: EvmAddress.fromBigInt(2n),
                srcToken: SolanaAddress.fromBigInt(3n),
                dstToken: EvmAddress.fromBigInt(4n), // address on dst chain
                srcAmount: parseEther('1'),
                minDstAmount: parseUnits('1000', 6)
            },
            {
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: NetworkEnum.ETHEREUM,
                srcSafetyDeposit: 1000n,
                dstSafetyDeposit: 1000n,
                timeLocks: TimeLocks.fromDurations({
                    srcFinalityLock: 10n,
                    srcPrivateWithdrawal: 200n,
                    srcPublicWithdrawal: 100n,
                    srcPrivateCancellation: 100n,
                    dstFinalityLock: 10n,
                    dstPrivateWithdrawal: 100n,
                    dstPublicWithdrawal: 100n
                }),
                hashLock: HashLock.forSingleFill(
                    '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
                )
            },
            {
                auction: AuctionDetails.noAuction(120n, 1234n)
            },
            {
                allowMultipleFills: false,
                salt: 123n
            }
        )

        const ix = SvmSrcEscrowFactory.DEFAULT.createOrder(order, {
            srcTokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
        })

        expect(ix).toMatchSnapshot()
    })

    it('should generate create instruction from order with bump bigger than u16', () => {
        const order = SvmCrossChainOrder.new(
            // 1 WETH [solana] -> 1000 USDC [ethereum]
            {
                maker: SolanaAddress.fromBigInt(1n),
                receiver: EvmAddress.fromBigInt(2n),
                srcToken: SolanaAddress.fromBigInt(3n),
                dstToken: EvmAddress.fromBigInt(4n), // address on dst chain
                srcAmount: parseEther('1'),
                minDstAmount: parseUnits('1000', 6)
            },
            {
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: NetworkEnum.ETHEREUM,
                srcSafetyDeposit: 1000n,
                dstSafetyDeposit: 1000n,
                timeLocks: TimeLocks.fromDurations({
                    srcFinalityLock: 10n,
                    srcPrivateWithdrawal: 200n,
                    srcPublicWithdrawal: 100n,
                    srcPrivateCancellation: 100n,
                    dstFinalityLock: 10n,
                    dstPrivateWithdrawal: 100n,
                    dstPublicWithdrawal: 100n
                }),
                hashLock: HashLock.forSingleFill(
                    '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
                )
            },
            {
                auction: new AuctionDetails({
                    duration: 120n,
                    initialRateBump: 70_000,
                    points: [],
                    startTime: 1234n
                })
            },
            {
                allowMultipleFills: false,
                salt: 123n
            }
        )

        const ix = SvmSrcEscrowFactory.DEFAULT.createOrder(order, {
            srcTokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
        })

        expect(ix.data.length).toBeGreaterThan(0)
    })

    it('should parse create instruction', async () => {
        const auction = AuctionDetails.noAuction(120n, 1752739636n)

        const order = SvmCrossChainOrder.new(
            // 1 WETH [solana] -> 1000 USDC [ethereum]
            {
                maker: SolanaAddress.fromBigInt(1n),
                receiver: EvmAddress.fromBigInt(2n),
                srcToken: SolanaAddress.fromBigInt(3n),
                dstToken: EvmAddress.fromBigInt(4n), // address on dst chain
                srcAmount: parseEther('1'),
                minDstAmount: parseUnits('1000', 6)
            },
            {
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: NetworkEnum.ETHEREUM,
                srcSafetyDeposit: 1000n,
                dstSafetyDeposit: 2000n,
                timeLocks: TimeLocks.fromDurations({
                    srcFinalityLock: 10n,
                    srcPrivateWithdrawal: 200n,
                    srcPublicWithdrawal: 100n,
                    srcPrivateCancellation: 150n,
                    dstFinalityLock: 20n,
                    dstPrivateWithdrawal: 300n,
                    dstPublicWithdrawal: 400n
                }),
                hashLock: HashLock.forSingleFill(
                    '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
                )
            },
            {
                auction
            },
            {
                allowMultipleFills: false,
                salt: 0x535n,
                orderExpirationDelay: 24n,
                resolverCancellationConfig:
                    ResolverCancellationConfig.ALMOST_ZERO,
                source: 'sdk'
            }
        )

        const ix = SvmSrcEscrowFactory.DEFAULT.createOrder(order, {
            srcTokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
        })

        const parsedIx = SvmSrcEscrowFactory.parseCreateInstruction(ix)

        const reCreatedOrder = SvmCrossChainOrder.fromContractOrder(
            parsedIx,
            auction
        )

        expect(order.toJSON()).toEqual(reCreatedOrder.toJSON())
        expect(order.getOrderHash(NetworkEnum.SOLANA)).toEqual(
            reCreatedOrder.getOrderHash(NetworkEnum.SOLANA)
        )
    })

    it('should generate withdrawPublic instruction', () => {
        const immutables = Immutables.new({
            orderHash: Buffer.from(
                '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
                'hex'
            ),
            hashLock: HashLock.forSingleFill(
                '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
            ),
            taker: SolanaAddress.fromBigInt(100n),
            token: SolanaAddress.fromBigInt(200n),
            maker: SolanaAddress.fromBigInt(300n),
            amount: parseEther('1'),
            safetyDeposit: 1000n,
            timeLocks: TimeLocks.fromDurations({
                srcFinalityLock: 10n,
                srcPrivateWithdrawal: 200n,
                srcPublicWithdrawal: 100n,
                srcPrivateCancellation: 100n,
                dstFinalityLock: 10n,
                dstPrivateWithdrawal: 100n,
                dstPublicWithdrawal: 100n
            })
        })

        const secret = Buffer.from(
            '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
            'hex'
        )
        const resolver = SolanaAddress.fromBigInt(400n)

        const ix = SvmSrcEscrowFactory.DEFAULT.withdrawPublic(
            immutables,
            secret,
            resolver,
            {
                tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
            }
        )

        expect(ix).toMatchSnapshot()
    })

    it('should parse create escrow instruction', async () => {
        const auction = new AuctionDetails({
            startTime: 1752739636n,
            initialRateBump: 100_000,
            duration: 120n,
            points: [
                {delay: 12, coefficient: 80_000},
                {delay: 24, coefficient: 50_000}
            ]
        })

        const secretHashes = [
            '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1',
            '0x1a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f2',
            '0x3a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f4',
            '0x5a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f6'
        ]

        const merkleLeafs =
            HashLock.getMerkleLeavesFromSecretHashes(secretHashes)

        const hashLock = HashLock.forMultipleFills(merkleLeafs)

        const order = SvmCrossChainOrder.new(
            // 1 WETH [solana] -> 1000 USDC [ethereum]
            {
                maker: SolanaAddress.fromBigInt(1n),
                receiver: EvmAddress.fromBigInt(2n),
                srcToken: SolanaAddress.fromBigInt(3n),
                dstToken: EvmAddress.fromBigInt(4n), // address on dst chain
                srcAmount: parseEther('1'),
                minDstAmount: parseUnits('1000', 6)
            },
            {
                srcChainId: NetworkEnum.SOLANA,
                dstChainId: NetworkEnum.ETHEREUM,
                srcSafetyDeposit: 1000n,
                dstSafetyDeposit: 2000n,
                timeLocks: TimeLocks.fromDurations({
                    srcFinalityLock: 10n,
                    srcPrivateWithdrawal: 200n,
                    srcPublicWithdrawal: 100n,
                    srcPrivateCancellation: 150n,
                    dstFinalityLock: 20n,
                    dstPrivateWithdrawal: 300n,
                    dstPublicWithdrawal: 400n
                }),
                hashLock
            },
            {
                auction
            },
            {
                allowMultipleFills: true,
                salt: 0x535n,
                orderExpirationDelay: 24n,
                resolverCancellationConfig:
                    ResolverCancellationConfig.ALMOST_ZERO,
                source: 'sdk'
            }
        )

        const taker = new SolanaAddress(
            'HjwhnenGQ2t24yVcDJ6uhXixY47UzsypsQaFuPZZJh6E'
        )

        const fillAmount = order.makingAmount / 2n

        const ix = SvmSrcEscrowFactory.DEFAULT.createEscrow(
            order.toSrcImmutables(
                NetworkEnum.SOLANA,
                taker,
                fillAmount,
                HashLock.fromString(secretHashes[1])
            ),
            auction,
            {
                tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID,
                merkleProof: {
                    proof: HashLock.getProof(merkleLeafs, 1),
                    idx: 1,
                    secretHash: bufferFromHex(secretHashes[1])
                }
            }
        )

        const parsedIx = SvmSrcEscrowFactory.parseCreateEscrowInstruction(ix)

        expect(parsedIx.amount).toEqual(fillAmount)
        expect(parsedIx.dutchAuctionData.toJSON()).toEqual(auction.toJSON())
        expect(parsedIx.merkleProof!.index).toEqual(1)
        expect(parsedIx.merkleProof!.hashedSecret).toEqual(secretHashes[1])
        expect(parsedIx.merkleProof!.proof).toEqual([
            '0x7a9cfaa16d89b92cb2e8e22fdb6042ec9439f3c3c8ca8565cbe3cd237f2d9e2c',
            '0x2f4f2437bc342f8ca7ce9889040d05d092a3ce902daa0b51c2b896c9a207dcd9'
        ])
    })

    it('should parse withdraw instruction', async () => {
        const immutables = Immutables.fromJSON<SolanaAddress>({
            orderHash:
                '0x62b5cf375b2e813bf2a0f33112712601d5aab04e598701f0bab2b6c5e9fa8a76',
            hashlock:
                '0x1a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f2',
            maker: '0x0000000000000000000000000000000000000000000000000000000000000001',
            taker: '0xf8bb3ce975e1ae20ccc5bd1e775828b2f811c617cafc6e4182d84c290a09f0f7',
            token: '0x0000000000000000000000000000000000000000000000000000000000000003',
            amount: '500000000000000000',
            safetyDeposit: '1000',
            timelocks:
                '4519513249946090673914462965909562690094454064409420748554250'
        })

        const secret = randomBytes(32)

        const ix = SvmSrcEscrowFactory.DEFAULT.withdrawPrivate(
            immutables,
            secret,
            {
                tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
            }
        )

        const parsed = SvmSrcEscrowFactory.parsePrivateWithdrawInstruction(ix)
        expect(parsed.secret).toEqual('0x' + secret.toString('hex'))
    })

    it('should parse publicWithdraw instruction', async () => {
        const immutables = Immutables.fromJSON<SolanaAddress>({
            orderHash:
                '0x62b5cf375b2e813bf2a0f33112712601d5aab04e598701f0bab2b6c5e9fa8a76',
            hashlock:
                '0x1a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f2',
            maker: '0x0000000000000000000000000000000000000000000000000000000000000001',
            taker: '0xf8bb3ce975e1ae20ccc5bd1e775828b2f811c617cafc6e4182d84c290a09f0f7',
            token: '0x0000000000000000000000000000000000000000000000000000000000000003',
            amount: '500000000000000000',
            safetyDeposit: '1000',
            timelocks:
                '4519513249946090673914462965909562690094454064409420748554250'
        })

        const secret = randomBytes(32)

        const ix = SvmSrcEscrowFactory.DEFAULT.withdrawPublic(
            immutables,
            secret,
            immutables.taker,
            {
                tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
            }
        )

        const parsed = SvmSrcEscrowFactory.parsePublicWithdrawInstruction(ix)
        expect(parsed.secret).toEqual('0x' + secret.toString('hex'))
    })

    it('should generate cancelPrivate instruction', () => {
        const immutables = Immutables.new({
            orderHash: Buffer.from(
                '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
                'hex'
            ),
            hashLock: HashLock.forSingleFill(
                '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
            ),
            taker: SolanaAddress.fromBigInt(100n),
            token: SolanaAddress.fromBigInt(200n),
            maker: SolanaAddress.fromBigInt(300n),
            amount: parseEther('1'),
            safetyDeposit: 1000n,
            timeLocks: TimeLocks.fromDurations({
                srcFinalityLock: 10n,
                srcPrivateWithdrawal: 200n,
                srcPublicWithdrawal: 100n,
                srcPrivateCancellation: 100n,
                dstFinalityLock: 10n,
                dstPrivateWithdrawal: 100n,
                dstPublicWithdrawal: 100n
            })
        })

        const ix = SvmSrcEscrowFactory.DEFAULT.cancelPrivate(immutables, {
            tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
        })

        expect(ix).toMatchSnapshot()
    })

    it('should generate cancelPublic instruction', () => {
        const immutables = Immutables.new({
            orderHash: Buffer.from(
                '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef',
                'hex'
            ),
            hashLock: HashLock.forSingleFill(
                '0x4a52dc502242a54e1d3a609cb31e0160a504d9a26467fcf9a52b7a79060ef8f1'
            ),
            taker: SolanaAddress.fromBigInt(100n),
            token: SolanaAddress.fromBigInt(200n),
            maker: SolanaAddress.fromBigInt(300n),
            amount: parseEther('1'),
            safetyDeposit: 1000n,
            timeLocks: TimeLocks.fromDurations({
                srcFinalityLock: 10n,
                srcPrivateWithdrawal: 200n,
                srcPublicWithdrawal: 100n,
                srcPrivateCancellation: 100n,
                dstFinalityLock: 10n,
                dstPrivateWithdrawal: 100n,
                dstPublicWithdrawal: 100n
            })
        })

        const payer = SolanaAddress.fromBigInt(400n)

        const ix = SvmSrcEscrowFactory.DEFAULT.cancelPublic(immutables, payer, {
            tokenProgramId: SolanaAddress.TOKEN_2022_PROGRAM_ID
        })

        expect(ix).toMatchSnapshot()
    })
})
