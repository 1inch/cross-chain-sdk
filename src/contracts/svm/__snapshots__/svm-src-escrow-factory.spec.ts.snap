// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SVM Escrow src factory should generate cancelPrivate instruction 1`] = `
{
  "accounts": [
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "11111111111111111111111111111112j",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111116B",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111114T",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "3pf4BAMrQPDhtY3xPfnjYdiViRmEVZmbBUHSG1UTjbiw",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "DCfwqvbJNpzmK7dTRwprUxGpGUXaZwYpUDuLMHkieUGf",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "14DvVcGoPg1Wh4fVbDcq9NHBxtMCXkqi7PczedJXFvFQ",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0x9ccb36b326482115",
  "programId": "2g4JDRMD7G3dK1PHmCnDAycKzd6e5sdhxqGBbs264zwz",
}
`;

exports[`SVM Escrow src factory should generate create instruction from order 1`] = `
{
  "accounts": [
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "11111111111111111111111111111112",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111114",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "ExSWT3R6qMY3L9ZYtpmMpyF5aX7AsJVeaMuhEyTBodLG",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "7HaPZKcfHrCJEQLZ8PUmg7YCfCcjawWg3rw8soDFW6DM",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "FyHBiQBaN5FFGYkWxdFzmtFiHn4k8hBvBkc6dc52JSqr",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "SysvarRent111111111111111111111111111111111",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0x181ec828051c077711f8dd293e31c96da8ed3c186011fd0eb9036cce32f2e08c1c5fd5001b737907000064a7b3b6e00de8030000000000000a000000d2000000360100009a0100000a0000006e000000d200000000000000560500000000ca9a3b00000000000000000000000000000000000000000000000000000000f2225665649874b9106f6b2867062e48a345a9c98cf31836497a594465e851f4010000000000000001000000007b000000d310e2140100000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000004e8030000000000000000000000000000",
  "programId": "2g4JDRMD7G3dK1PHmCnDAycKzd6e5sdhxqGBbs264zwz",
}
`;

exports[`SVM Escrow src factory should generate withdrawPublic instruction 1`] = `
{
  "accounts": [
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "11111111111111111111111111111112j",
    },
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": "1111111111111111111111111111117u",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "Aikth6gKPGZbWbgkZSZumpXpBjTQLq2UUMccZ2CRhtSQ",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111114T",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "3pf4BAMrQPDhtY3xPfnjYdiViRmEVZmbBUHSG1UTjbiw",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "DCfwqvbJNpzmK7dTRwprUxGpGUXaZwYpUDuLMHkieUGf",
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": "cKjt7npJFbBC1Lu5FvKmgnVCKYYrBsmCkAWUnSNHYN8",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": "11111111111111111111111111111111",
    },
  ],
  "data": "0x9839f0c05223960b0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
  "programId": "2g4JDRMD7G3dK1PHmCnDAycKzd6e5sdhxqGBbs264zwz",
}
`;
